src/test/unit/callbacks/json_writer_test.cpp

```diff
@@ -0,0 +1,142 @@
+#include <gtest/gtest.h>
+#include <boost/lexical_cast.hpp>
+#include <stan/callbacks/json_writer.hpp>
+
+struct deleter_noop {
+  template <typename T>
+  constexpr void operator()(T* arg) const {}
+};
+class StanInterfaceCallbacksJsonWriter : public ::testing::Test {
+ public:
+  StanInterfaceCallbacksJsonWriter()
+      : ss(), writer(std::unique_ptr<std::stringstream, deleter_noop>(&ss)) {}
+
+  void SetUp() {
+    ss.str(std::string());
+    ss.clear();
+  }
+
+  void TearDown() { writer.reset(); }
+
+  std::stringstream ss;
+  stan::callbacks::json_writer<std::stringstream, deleter_noop> writer;
+};
+
+TEST_F(StanInterfaceCallbacksJsonWriter, begin_record_end) {
+  writer.begin_record();
+  writer.end_record();
+  EXPECT_EQ("{}", ss.str());
+}
+
+TEST_F(StanInterfaceCallbacksJsonWriter, write_double_vector) {
+  std::string key("key");
+  const int N = 5;
+  std::vector<double> x;
+  for (int n = 0; n < N; ++n)
+    x.push_back(n);
+
+  writer.write(key, x);
+  EXPECT_EQ("\"key\" : [ 0, 1, 2, 3, 4 ]", ss.str());
+
+  writer.write(key, x);
+  EXPECT_EQ(
+      "\"key\" : [ 0, 1, 2, 3, 4 ]"
+      ", \"key\" : [ 0, 1, 2, 3, 4 ]",
+      ss.str());
+
+  writer.write(key, x);
+  EXPECT_EQ(
+      "\"key\" : [ 0, 1, 2, 3, 4 ]"
+      ", \"key\" : [ 0, 1, 2, 3, 4 ]"
+      ", \"key\" : [ 0, 1, 2, 3, 4 ]",
+      ss.str());
+}
+
+TEST_F(StanInterfaceCallbacksJsonWriter, single_member) {
+  std::string key("key");
+  std::string value("value");
+  writer.begin_record();
+  writer.write(key, value);
+  writer.end_record();
+  EXPECT_EQ("{\"key\" : \"value\"}", ss.str());
+}
+
+TEST_F(StanInterfaceCallbacksJsonWriter, more_members) {
+  std::string key("key");
+  std::string value("value");
+  writer.begin_record();
+  writer.write(key, value);
+  writer.write(key, value);
+
+  EXPECT_EQ(
+      "{\"key\" : \"value\""
+      ", \"key\" : \"value\"",
+      ss.str());
+
+  writer.write(key, value);
+  writer.end_record();
+  EXPECT_EQ(
+      "{\"key\" : \"value\""
+      ", \"key\" : \"value\""
+      ", \"key\" : \"value\"}",
+      ss.str());
+}
+
+TEST_F(StanInterfaceCallbacksJsonWriter, write_double_vector_precision2) {
+  ss << std::setprecision(2);
+  std::string key("key");
+  const int N = 5;
+  std::vector<double> x{1.23456789, 2.3456789, 3.45678910, 4.567890123};
+  writer.write(key, x);
+  EXPECT_EQ("\"key\" : [ 1.2, 2.3, 3.5, 4.6 ]", ss.str());
+}
+
+TEST_F(StanInterfaceCallbacksJsonWriter, write_double_vector_nan_inf) {
+  std::string key("key");
+  std::vector<double> x;
+  x.push_back(std::numeric_limits<double>::quiet_NaN());
+  x.push_back(std::numeric_limits<double>::infinity());
+  x.push_back(-std::numeric_limits<double>::infinity());
+  writer.write(key, x);
+  EXPECT_EQ("\"key\" : [ nan, inf, -inf ]", ss.str());
```

Comment:
The output here is not something which is readable by our own JSON parser for inputs. Tested with this model:

```stan
data {
   array[3] real key;
}
transformed data {
   print(key);
}
```

output:
```
Error in JSON parsing 
at offset 12: 
Invalid value.
```

Outputting the correct values (especially if they appear inside of an Eigen matrix or something) will be a bit tedious, but it is important if we ever want to consume the output of these files.

Reply:
fixed.